### PR TITLE
adding visual window implementation

### PR DESCRIPTION
**Pull request by hfazai**

### CODE REVIEW

File: galite-core/src/main/kotlin/org/kopi/galite/db/Connection.kt [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+      throw InconsistencyException("user id not yet determined")
+    }
+    return user
+  }
+
+  /**
+   * Retrieves the user ID of the current user
+   */
+  private fun setUserID() {
+    TODO()
```
[the remark this answers is not in the document]
Comment (hfazai) [reply galite-core/src/main/kotlin/org/kopi/galite/db/Connection.kt:97]: Yes, we will do that

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 2]
```diff
@@ -18,8 +18,615 @@
...
+        }
+      }
+      i -= 2
+    }
+    return false
+  }
+
+  fun getTitle(): String {
+    return title + if (extraTitle != null) " $extraTitle" else ""
+  }
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:235]: change this to inline function

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 3]
```diff
@@ -18,8 +18,615 @@
...
+    if (actors.isEmpty()) {
+      actors = ArrayList(actorDefs.size)
+    }
+    actors.addAll(actorDefs)
+  }
+
+  fun getActor(at: Int): VActor {
+    // "+1" because of the f12-Actor
+    return actors[at + 1]
+  }
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:267]: change this to inline function

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 4]
```diff
@@ -18,8 +18,615 @@
...
+
+  fun getActors(): ArrayList<VActor> {
+    return actors
+  }
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:271]: change this to inline function

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 5]
```diff
@@ -18,8 +18,615 @@
...
+      it.localize(manager)
+    }
+  }
+
+  /**
+   * close model if allowed
+   */
+  fun willClose(code: Int) {
+    close(code)
+  }
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:300]: change this to inline function

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 6]
```diff
@@ -18,8 +18,615 @@
...
+      }
+    }
+  }
+
+  //----------------------------------------------------------------------
+  // VMODEL IMPLEMENTATION
+  //----------------------------------------------------------------------
+  override fun getDisplay(): UWindow? {
+    return display
+  }
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:319]: change this to inline function

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 7]
```diff
@@ -18,8 +18,615 @@
...
+
+  // ----------------------------------------------------------------------
+  // UTILS
+  // ----------------------------------------------------------------------
+  /**
+   * setInformationText
+   */
+  fun setInformationText(text: String) {
+    display?.setInformationText(text)
+  }
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:335]: change this to inline function

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 8]
```diff
@@ -18,8 +18,615 @@
...
+    for (i in listeners.size - 2 downTo 0 step 2) {
+      if (listeners[i] === ProgressDialogListener::class.java) {
+        (listeners[i + 1] as ProgressDialogListener).setProgressDialog(message, currentJob)
+      }
+    }
+  }
+
+  fun setTotalJobs(totalJobs: Int) {
+    getDisplay()?.setTotalJobs(totalJobs)
+  }
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:351]: change this to inline function

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 9]
```diff
@@ -18,8 +18,615 @@
...
+
+  fun setCurrentJob(currentJob: Int) {
+    getDisplay()?.setCurrentJob(currentJob)
+  }
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:355]: change this to inline function

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 10]
```diff
@@ -18,8 +18,615 @@
...
+    for (i in listeners.size - 2 downTo 0 step 2) {
+      if (listeners[i] === ProgressDialogListener::class.java) {
+        (listeners[i + 1] as ProgressDialogListener).unsetProgressDialog()
+      }
+    }
+  }
+
+  fun updateWaitDialogMessage(message: String) {
+    getDisplay()?.updateWaitDialogMessage(message)
+  }
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:368]: change this to inline function

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 11]
```diff
@@ -18,8 +18,615 @@
...
+
+  /**
+   * Asks the user, if she/he wants to retry the exception
+   *
+   * @return true, if the transaction should be retried.
+   */
+  fun retryProtected(): Boolean {
+    return ask(MessageCode.getMessage("VIS-00039"))
+  }
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:466]: change this to inline function

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 12]
```diff
@@ -18,8 +18,615 @@
...
+
+  /**
+   * return wether this object handle a transaction at this time
+   */
+  fun inTransaction(): Boolean {
+    return isProtected
+  }
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:473]: change this to inline function

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 13]
```diff
@@ -18,8 +18,615 @@
...
+
+  /**
+   * Returns the current user name
+   */
+  open fun getUserName(): String? {
+    return dBContext.defaultConnection.userName
+  }
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:480]: change this to inline function

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 14]
```diff
@@ -18,8 +18,615 @@
...
+
+  /**
+   * Returns the user ID
+   */
+  open fun getUserID(): Int {
+    return dBContext.defaultConnection.getUserID()
+  }
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:487]: change this to inline function

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 15]
```diff
@@ -18,8 +18,615 @@
...
+  /**
+   * Formats the message having the given identifier from the given source.
+   *
+   * @param     ident the message identifier
+   * @param     param message parameter
+   * @return    the requested message
+   */
+  protected fun formatMessage(ident: String, param: Any? = null): String? {
+    return formatMessage(ident, param, null)
+  }
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:501]: change this to inline function

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 16]
```diff
@@ -18,8 +18,615 @@
...
+   * Formats the message having the given identifier from the given source.
+   *
+   * @param     ident  the message identifier
+   * @param     param1 the first message parameter
+   * @param     param1 the second message parameter
+   * @return    the requested message
+   */
+  protected fun formatMessage(ident: String, param1: Any?, param2: Any?): String? {
+    return formatMessage(ident, arrayOf(param1, param2))
+  }
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:513]: change this to inline function

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 17]
```diff
@@ -18,8 +18,615 @@
...
+    modelListener.remove(VActionListener::class.java, al)
+  }
+
+  fun addModelCloseListener(mcl: ModelCloseListener) {
+    modelListener.add(ModelCloseListener::class.java, mcl)
+  }
+
+  fun removeModelCloseListener(mcl: ModelCloseListener) {
+    modelListener.remove(ModelCloseListener::class.java, mcl)
+  }
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:579]: can you change those methods to inline methods?

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 18]
```diff
@@ -18,8 +18,615 @@
...
+  // --------------------------------------------------------------------
+  /**
+   * Adds a listener to the list that's notified each time a production
+   * of the report file occurs.
+   *
+   * @param l The FileProductionListener
+   */
+  fun addFileProductionListener(l: FileProductionListener) {
+    listenerList.add(FileProductionListener::class.java, l)
+  }
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:592]: inline method

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt [hunk 19]
```diff
@@ -18,8 +18,615 @@
...
+
+  /**
+   * Removes a listener from the list that's notified each time a
+   * production of the report file occurs.
+   *
+   * @param l The FileProductionListener
+   */
+  fun removeReportListener(l: FileProductionListener) {
+    listenerList.remove(FileProductionListener::class.java, l)
+  }
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VWindow.kt:602]: inline method

**Review by h-haddad — APPROVED** [review]

**Review by h-haddad — APPROVED** [review]

**Pull request merged**